### PR TITLE
fix(sdk): Add outputDefinitions of root pipeline to compiled components IR YAML file

### PR DESCRIPTION
**Pull request by JOCSTAA**
Enables the output_definitions of a component to be surfaced to the root pipeline output_definitions in the IR YAML file when compiling a component as a pipeline.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler_test.py [hunk 1]
```diff
@@ -3080,5 +3080,23 @@ def my_pipeline(sample_input1: str = 'message'):
...
 
+class TestOutputDefinitionsPresentWhenCompilingComponents(unittest.TestCase):
+
+    def test_basic_permitted(self):
+
+        @dsl.component
+        def comp(message: str) -> str:
+            return message
+
+        with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_test.py:3091]: It's possible to assert on `identity.pipeline_spec.root.output_definitions` directly, which avoids the compile/load steps

File: sdk/python/kfp/components/structures.py [hunk 2]
```diff
@@ -931,12 +942,11 @@ def to_pipeline_spec(self) -> pipeline_spec_pb2.PipelineSpec:
...
             self)
-        has_inputs = bool(
-            len(component_spec_proto.input_definitions.artifacts) +
-            len(component_spec_proto.input_definitions.parameters))
-        if has_inputs:
-            pipeline_spec.root.input_definitions.CopyFrom(
-                component_spec_proto.input_definitions)
+        pipeline_spec.root.CopyFrom(component_spec_proto)
+
+        if pipeline_outputs:
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/structures.py:947]: Is `if pipeline_outputs` necessary? I think the `_builder_dag_outputs` generalizes to the case of an empty `dag_outputs` dict.

File: sdk/python/kfp/components/structures.py [hunk 3]
```diff
@@ -919,6 +919,17 @@ def to_pipeline_spec(self) -> pipeline_spec_pb2.PipelineSpec:
         pipeline_name = self.name
         task_group = group
 
+        pipeline_outputs = {}
+        pipeline_inputs = self.inputs or {}
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/structures.py:923]: I'm not sure this line is doing anything. I think this is duplicative of the [line above](https://github.com/kubeflow/pipelines/blob/ef5d67252949b85f3726cbfe905d2aa14ac26b1f/sdk/python/kfp/components/structures.py#L903).

File: sdk/python/test_data/components/component_with_pip_install.yaml [hunk 4]
```diff
@@ -19,7 +19,7 @@ deploymentSpec:
           \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
           \ python3 -m pip install --quiet     --no-warn-script-location --index-url\
           \ https://pypi.org/simple --trusted-host https://pypi.org/simple 'yapf'\
-          \ 'kfp==2.0.0-beta.8' && \"$0\" \"$@\"\n"
+          \ 'kfp==2.0.0-beta.12' && \"$0\" \"$@\"\n"
```
Comment (connor-mccarthy) [inline sdk/python/test_data/components/component_with_pip_install.yaml:22]: Can we avoid the diff for files that aren't affected by this fix? This diff just a version bump.

File: sdk/python/test_data/components/output_metrics.yaml [hunk 5]
```diff
@@ -2,6 +2,7 @@
 # Name: output-metrics
 # Description: Dummy component that outputs metrics with a random accuracy.
 # Outputs:
+#    metrics: system.Metrics
 #    output-metrics-metrics: system.Metrics
```
Comment (connor-mccarthy) [inline sdk/python/test_data/components/output_metrics.yaml:6]: Not part of this PR, but do you know what is going on here with `output-metrics-metrics`?

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]

File: sdk/python/test_data/components/output_metrics.yaml (hunk 5 above)
Comment (JOCSTAA) [reply sdk/python/test_data/components/output_metrics.yaml:6]: I have no idea why that is happening

File: sdk/python/test_data/components/component_with_pip_install.yaml (hunk 4 above)
Comment (JOCSTAA) [reply sdk/python/test_data/components/component_with_pip_install.yaml:22]: I think all those changed when i auto generated the golden snapshots

File: sdk/python/test_data/components/container_no_input.yaml [hunk 6]
```diff
@@ -24,4 +24,4 @@ root:
         taskInfo:
           name: container-no-input
 schemaVersion: 2.1.0
-sdkVersion: kfp-2.0.0-beta.8
+sdkVersion: kfp-2.0.0-beta.12
```
Comment (connor-mccarthy) [inline sdk/python/test_data/components/container_no_input.yaml:27]: nit: can remove this diff

**Comment by connor-mccarthy** [comment]
/lgtm

Thanks, @JOCSTAA!
